flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/source/IcebergSource.java

```diff
@@ -429,6 +440,20 @@ public Builder<T> setAll(Map<String, String> properties) {
       return this;
     }
 
+    /**
+     * Sets the {@link IcebergWatermarkExtractor} to retrieve the split watermark before emitting
+     * the records for a given split. The {@link
+     * IcebergWatermarkExtractor#extractWatermark(IcebergSourceSplit)} is also used for ordering the
+     * splits for read.
+     */
+    public Builder<T> watermarkExtractor(IcebergWatermarkExtractor<T> newWatermarkExtractor) {
```

Comment:
nit: Instead of requiring the user to provide watermark extractor here and then creating the assignerFactory, a better approach would be to allow the user to create the assignerFactory with the watermark extractor. This would enable the decoupling of the iceberg source from watermark extraction. As a result, the code could be simplified, and some of the preconditions could be removed.

Reply:
I can see where @sundargates came from. Original API exposes `SplitAssignerFactory` in the `IcebergSource$Builder`. the intention is to allow users to switch to different assigners (like ordered, locality aware assigners).

@sundargates with the new approach, readers need to extract split level watermark and emit the watermark values. Hence just passing timestamp extractor to assigner is not enough.

But I also realized that maybe exposing `SplitAssignerFactory` is not very intuitive to end users. In another comment, I suggested `sourceTimestampField (or) sourceWatermarkField(String fieldName, TimeUnit unit)`

Reply:
Gone for:
```
    IcebergSource.<RowData>builder()
        .tableLoader(sourceTableResource.tableLoader())
        .watermarkColumn("ts")
```